[2단계 - 로또(자동)] 렉스(오성원) 미션 제출합니다.

Seongwon97 opened the pull request:
안녕하세요 코니😁

지난 1단계 추가로 주신 피드백을 반영 후 2단계 미션도 구현 완료하여 리뷰 요청드립니다.

이번 미션에는 다음과 같은 내용들에 신경을 써서 리펙터링&기능구현을 하였습니다.
1. 로또의 번호 저장 변수를 `List`에서 `Set`으로 변경하여 중복체크를 쉽게 하도록 수정하였습니다.
2. `utils`의 역할 이해 및 utils에 있던 메서드들을 올바른 위치로 이동시켰습니다.
3. 메서드들을 각각의 필요한 객체들에 위치시키려 노력했습니다.
4. 중복된 메서드를 재사용할 수 있도록 노력했습니다. (당첨번호와 수동 로또번호 검증에 `ManualLottoGenerator`를 사용) 

이번에도 잘 부탁드리겠습니다! 감사합니다.🙇🏻‍♂️

Seongwon97 committed:
refactor: 1단계 피드백 반영하여 리펙터링

- (Lottos) 생성자 코드 축약
- (OutputView) 구매 로또 출력 기능 수정
- (NumberTest) 메서드 이름과 DisplayName을 하는 일에 맞게 올바르게 수정
- (PrizeResultTest) 인스턴스 변수를 메서드 내부로 이동

Seongwon97 committed:
refactor: Lotto와 WinningNumbers의 변수 타입을 List에서 Set으로 변경

- Number를 저장하는 변수를 List에서 Set으로 변경하여 중복되지 않는 숫자만 저장하도록 변경

Seongwon97 committed:
refactor: 유효한 rank값을 반환하는 메서드의 위치를 변경

- FAIL을 제외한 유효한 Rank값들을 반환하는 메서드의 위치를
PrizeResult에서 Rank로 변경

Seongwon97 committed:
refactor: InputValidation 내의 검증 메서드들을 각각의 필요한 객체로 이동

- utils는 로직과 관계가 없고 범용적으로 사용될 수 있는 도구들이 위치하여야 하기에
로직과 관련이 있는 InputValidation의 검증 메서드들을 각각의 검증이 필요한 Model로 이동

Seongwon97 committed:
refactor: 파일 끝에 개행 추가

- 파일 끝에 개행 추가
- NumberTest파일 메서드의 DisplayName수정

Seongwon97 committed:
docs: 2단계 요구사항 정리 추가

Seongwon97 committed:
feat: 로또 수동 구매기능 추가

- 수동 구매 전략을 추가하여 구현
- 보유 돈을 체크하기 위해 Money객체를 생성

Seongwon97 committed:
refactor: 로또 넘버 생성 전략을 LottoGenerator로 이름 변경

- 프로그램적으로 생각했을 때 전략보다는 로또번호 생성이 더 좋은 이름일거 같아서
LottoGenerator, ManualLottoGenerator, AutomaticLottoGenerator로 이름 변경

Seongwon97 committed:
refactor: WinningNumber에서도 ManualLottoGenerator를 활용하여 번호 생성하도록 수정

- WinningNumber와 ManualLottoGenerator의 중복 코드를 줄이고자
ManualLottoGenerator의 메서드를 활용하여 당첨번호를 생성하도록 수정

Seongwon97 committed:
refactor: 에러 메시지 수정, 불필요한 상수제거, 변수에 final추가

- (Money) 에러 메시지 상수 이름 수정
- (Store) 사용하지 않는 상수 제거
- (WinningNumbers) bonus에 final추가

Seongwon97 committed:
refactor: 메서드의 final 규칙 통일

- 메서드의 매개변수는 primitive type과 String에만 final을 붙이고
그 외의 reference type에는 붙이지 않는다.

Seongwon97 committed:
refactor: 출력문 줄바꿈 조정

choihz commented on src/test/java/domain/NumberTest.java:
```diff
@@ -11,18 +11,18 @@
     @ParameterizedTest(name = "{index} {displayName} bonus={0}")
     @ValueSource(ints = {0, -1, 46})
-    void checkBonusOutRange_throwIllegalException(final int bonus) {
+    void checkNumberOutRange_throwIllegalException(final int bonus) {
```
아직도 `bonus`가 사용되고 있네요~

Seongwon97 replied:
로또 번호의 범위를 검사하는 테스트에 전혀 무관한 bonus라는 변수가 들어갔네요..

제거 완료했습니다!

choihz commented on src/test/java/domain/NumberTest.java:
```diff
@@ -11,18 +11,18 @@
 
     @Test
-    @DisplayName("같은 숫자를 가진 로또 넘버 객체는 동일하다.")
+    @DisplayName("같은 숫자를 가진 로또 넘버 객체는 동등하다.")
```
정리해 주신 블로그 글 잘 읽었습니다 👍👍

choihz commented on src/main/java/domain/generator/ManualLottoGenerator.java:
```diff
@@ -0,0 +1,63 @@
+public class ManualLottoGenerator implements LottoGenerator {
+
+    private static final int LOTTO_SIZE = 6;
+    private static final String LOTTO_NUMBER_DELIMITER = ", |,";
```
이 구분자는 콘솔 뷰에 한정된 것 아닐까요? 이 정보를 `ManualLottoGenerator`가 가지고 있어야 할까요?

Seongwon97 replied:
view에서 하는 일을 단순히 입력만 받도록 줄이다보니 이러한 문제가 발생했네요ㅠ

입력 형식을 변경할 경우 view가 아니라 domain을 변경해야하는 문제가 생겼네요.
숫자 구분자를 domain이 아닌 view에 위치하도록 변경해봤습니다😁

choihz commented on src/main/java/domain/generator/ManualLottoGenerator.java:
```diff
@@ -0,0 +1,63 @@
+        if (numbers.size() != LOTTO_SIZE) {
+            throw new IllegalArgumentException(ERROR_DUPLICATE_NUMBER);
+        }
+    }
```
사이즈 체크를 이 메서드에서도 하고 `checkNumOfBalls`에서도 하는데, 두 번 할 필요가 있는 걸까요? 또한 비슷한 로직이 `Lotto` 안에도 있는데 한 곳에서 이 작업을 할 수는 없을까요?

Seongwon97 replied:
리펙터링을하며 같은 일을 하는 중복된 메서드가 발생하였네요..
체크하지 못한 것 같습니다ㅠ

반복된 작업을 한번만 수행하도록 변경했습니다😁

choihz commented on src/main/java/domain/Store.java:
```diff
@@ -1,26 +1,64 @@
+import domain.generator.ManualLottoGenerator;
+import domain.generator.LottoGenerator;
 
 public class Store {
```
너무 추상적인 이름 같아요.

Seongwon97 replied:
헉..그런가요🥲

로또 구매가 가능한지 체크하고 구매를 하는 역할을 하여 store라는 이름을 지었었는데 추상적이로 보이는군요ㅠ

더욱 구체적인 LottoMachine은 어떻게 생각되시나요?? Store를 LottoMachine으로 변경해봤습니다.

choihz commented on src/main/java/domain/Store.java:
```diff
@@ -1,26 +1,64 @@
+        checkNegativeInteger(numOfManualLotto);
+
+        return numOfManualLotto;
+    }
```
이 메서드가 정말 **검증**하는 일만 하고 있나요?

Seongwon97 replied:
단순히 검증만 하는 것이 아니라 반환을 int로 변환하여 해주기에 해당 작업을 하는 메서드의 이름은 항상 고민을 하고 있습니다.

메서드 명을 현재는 `validateAndChangeToInt`로 변경을 해봤습니다.

변경된 메서드 명은 어떻게 생각하시나요? 저는 메서드의 내부 작업을 너무 많이 드러내는 것 같기는 한데..검증 뿐만 아니라 값 변환까지 해주는 것을 표현하여 이전의 이름보다는 더 좋은 것 같다고 생각됩니다👍🏻

choihz commented on src/main/java/domain/Lottos.java:
```diff
@@ -6,12 +6,12 @@
 
 public class Lottos {
 
+    private static final int LOTTO_PRICE = 1000;
```
이게 꼭 `Lottos`가 알아야 하는 정보일까요?

Seongwon97 replied:
로또의 가격은 `LottoMachine`에서만, 로또의 크기는 `Lotto`에서만 알아도 될 것으로 판단하여 각각의 클래스에 public 접근 제어자의 상수로 정의하였습니다. 다른 클래스에서 사용할 경우 `LottoMachine`와 `Lotto`에서 정의된 값을 사용하도록 변경해봤습니다!

choihz commented on src/main/java/domain/Lottos.java:
```diff
@@ -22,4 +22,8 @@ public PrizeResult prizeResult(WinningNumbers winningNumber) {
 
+    public int amountOfLottos() {
+        return lottos.size() * LOTTO_PRICE;
+    }
```
이름이 드러내는 역할과 실제로 하는 일이 다른 것 같아요.

Seongwon97 replied:
앗...그렇네요😅

메서드명을 `totalPurchasePrice`으로 변경했습니다!

choihz commented on src/main/java/domain/Money.java:
```diff
@@ -0,0 +1,44 @@
+        } catch (NumberFormatException e) {
+            throw new IllegalArgumentException(Money.ERROR_PRICE_NON_INTEGER);
+        }
+    }
```
이곳을 비롯하여 `WinningNumbers`, `ManualLottoGenerator` 등에서 사실은 비즈니스 로직과는 관련이 없는 이 처리가 반복되고 있는 것으로 보여요. 개선할 방법은 없을까요?

Seongwon97 replied:
view의 작업을 단순히 입력만 받아 controller-model로 넘겨주는 형태를 만들려다보니 반복되더라도 각각의 domain들에서 해당 작업을 진행하게 하였습니다.

하지만 다시 생각해보니 단순히 정수로 변환이 가능한가?를 체크하며 정수형으로 변환해주는 메서드는 각각의 domain들이 하는 비즈니스와는 거리가 멀어 해당 작업은 view에서 진행하여도 되지 않을까?하는 생각이 드네요.

해당 메서드들은 domain에서 제거하고 view로 이동하여 모든 입력에 대해 재사용을 하도록 수정해봤습니다.

> 수정을 하고 코멘트를 남기다보니 해당 메서드의 작업이 바로 범용적으로 쓰일 수 있는 `util`속성에 맞지 않나?라는 생각이 들어 util을 다시 생성하여 이동시켰습니다.

choihz commented on src/main/java/domain/Money.java:
```diff
@@ -0,0 +1,44 @@
+
+    public void purchaseLotto(int numOfLotto) {
+        money -= (LOTTO_PRICE * numOfLotto);
+    }
```
모르고 막 쓰다가는 `money` 값이 음수가 될 수도 있겠네요.

> 생성자로 들어오는 리스트에 숫자가 한 개 있다면, 숫자 하나로 로또를 생성하겠네요. 또는 리스트에 숫자가 여섯 개 들어오지만 같은 게 두 개 있다면, 숫자 다섯 개로 로또를 생성하겠네요. 그렇다면 로또 객체를 사용하는 입장에서, 이 객체가 항상 완전하다고 보장할 수 있을까요?

1단계에서 드렸던 위 코멘트와 비슷한 맥락에서 고민해 보시면 좋을 것 같아요.

Seongwon97 replied:
그렇네요.. money의 입장에서 봤을 때 0보다 적을 경우 예외를 발생시켜 0이상의 수만 존재하도록 보장하도록 수정하였습니다!!👍🏻

choihz commented on src/main/java/domain/WinningNumbers.java:
```diff
@@ -1,38 +1,41 @@
-    private static final int LOTTO_SIZE = 6;
-    private static final String ERROR_DUPLICATE_NUMBER = "로또 번호는 중복되면 안 됩니다.";
     private static final String ERROR_DUPLICATE_BONUS = "보너스 번호가 당첨 번호와 중복됩니다.";
+    private static final String ERROR_BALL_NON_INTEGER = "로또 번호는 숫자만 입력해줘야 합니다.";
```
이 에러 메시지가 `WinningNumbers`에 위치하는 게 적절할까요?

전체적으로 역할과 책임이 잘 나눠져 있는지 조금 더 고민해 보시면 좋을 것 같아요. 하나의 메서드는 하나의 역할만 하는지, 클래스는 명확한 이름을 가지며 그에 걸맞은 일을 하고 있는지, 불필요한 중복은 없는지, 기능 변경이 있을 때 수정이 필요한 곳을 바로 찾아갈 수 있는지, 관련 있는 코드가 한곳에 모여 있는지 등 체크 리스트를 만들어 코드 전체를 점검해 보아도 좋구요.

Seongwon97 replied:
해당 내용도 [앞의 코멘트](https://github.com/woowacourse/java-lotto/pull/471#discussion_r818973354) 를 수정하며 불필요한 중복들을 제거와 메서들의 위치 조정을 해봤습니다!!

좋은 코멘트 주셔서 감사합니당🙇🏻‍♂️

choihz reviewed (changes requested):
안녕하세요, 렉스!

2단계도 잘 구현해 주셨네요 👍 몇 가지 코멘트 남겼는데 확인 부탁드리고, 이외에도 더 궁금한 점이나 의견이 있으시다면 언제든 코멘트 또는 DM 주세요~

Seongwon97 committed:
refactor: 중복된 검증 제거, view와 관련있는 구분자를 domain에서 view로 이동

- 입력된 로또 번호의 수를 검증하는 중복된 메서드 제거
- view와 관련있는 구분자를 ManualLottoGenerator.java에서 제거 후 view로 이동

Seongwon97 committed:
refactor: store을 lottoMachine으로 이름 변경

Seongwon97 committed:
refactor: 로또 번호 범위를 검증하는 테스트의 매개변수 이름 변경

- 로또 번호의 범위를 체크하는 NumberTest.java의 테스트 메서드의
매개변수를 bonus에서 number로 변경

Seongwon97 committed:
refactor: 메서드, 변수의 이름 수정

메서드와 변수의 이름을 명확하고 통일성있게 수정
- (Lottos): amountOfLottos() -> totalPurchasePrice()
- 모든 파일의 price -> money

Seongwon97 committed:
refactor: 로또 가격을 LottoMachine에 public접근 제어자의 상수로 설정

- Lottos, Money에 있던 로또 가격을 제거후 LottoMachine에서만 갖는
로또 가격 상수의 값을 불러와서 사용하도록 변경

Seongwon97 committed:
refactor: 입력값이 음수인지 체크 후 int형으로 변환하는 메서드를 view로 이동

- Money, WinningNumber, ManualLottoGenerator등에서 공통으로 갖던
checkNonInteger()메서드를 view로 이동하여 재사용하도록 변경
- checkNonInteger()메서드의 이름을 하는 작업을 표현하여 splitAndChangeToInt()로 변경
- 불필요한 메서드들 제거

Seongwon97 committed:
refactor: view에 위치한 입력값 변환 및 검증 메서드를 util로 이동

- splitAndChangeToInt(), validateAndChangeToInt() 메서드를 view에서 util로 이동

Seongwon97 committed:
style: 코드 컨벤션 수정

Seongwon97 committed:
style: 코드 컨벤션 통일

Seongwon97 committed:
feat: 돈이 0보다 적을 때 발생하는 예외 추가

Seongwon97 commented:
안녕하세요 코니!
보내주신 피드백들을 받고 몇몇 내용을 수정해봤어요!

수정을 하며 중복된 코드의 제거, mvc 및 각각의 객체들이 올바른 작업들을 갖고 있는지에 대하여 신경을 쓴것 같습니다.

수정을 하며 이것이 올바른 위치치인가..?하는 몇가지 의문점들이 있었습니다.
1. 입력된 로또 번호에 대해 6개가 맞는지와 중복체크를 하는 메서드의 위치에 대한 고민
해당 메서드의 경우는 `generator`에 위치시는 것이 좋은가?아니면 `Lotto`와 `WinningNumber`에 위치시키는 것이 좋은가에 대해 고민이 있었습니다. 
제 생각으로는 해당 검증 메서드는 둘 중 어디에 위치시켜도 괜찮지만 중복되더라도 `Lotto`와 `WinningNumber`에 위치시키는 것이 객체들이 올바른 값들을 갖는 것을 보증할 수 있을것이라 생각하여 해당 위치에 뒀습니다.
코니는 해당 메서드의 더 좋은 위치는 어디라고 생각하시나요??

2. 입력값에 대해 숫자로 변경가능한지 체크하고 정수로 변경하는 메서드, 그리고 로또본호 입력값에 대하여 split으로 분리를 하는 메서드는 프로그램 내에서 범용적으로 사용될 수 있다고 생각하여 util 클래스를 생성하여 해당 클래스 내에 위치시켰습니다.
해당 메서드들은 util성격에 맞다고 볼 수 있을까요??

감사합니다🙇🏻‍♂️

choihz commented on src/main/java/domain/WinningNumbers.java:
```diff
@@ -1,38 +1,42 @@
     private static final String ERROR_DUPLICATE_BONUS = "보너스 번호가 당첨 번호와 중복됩니다.";
+    private static final String ERROR_DUPLICATE_NUMBER = "로또 번호는 중복되지 않은 6자리 숫자여야 합니다.";
 
     private final Set<Number> winningNumbers;
```
`WinningNumbers`가 `Set<Number>`를 직접 가지고 있기보다는 `Lotto`를 가지고 있으면 불필요한 중복이 줄어들고 코드도 조금 더 명확해질 수 있지 않을까요? 지금은 둘이 비슷한 로직을 각각 가지고 있는 것 같아요. 1번 질문에 대한 답도 될 수 있겠네요.

Seongwon97 replied:
해당 내용을 이전에는 Lotto객체 내부의 당첨 등수를 확인하는 메서드가 WinningNumbers에서는 필요가 없다고 생각하여 구현하지 못하였습니다. 하지만 오늘 크루인 이프와 만나 몇몇 대화를 나누던 중 당첨 등수를 확인하는 메서드는 Lotto객체와  Rank객체에서 행위를 분리하는 것 보다는 Rank객체 내에서 당첨 등수를 확인하는 행위를 모두 갖는게 좋다는 생각을 하여 `calculateRank`메서드를 Rank로 옮긴 후, WinningNumbers에서 Lotto를 사용하도록 수정해봤습니다😁

choihz commented on src/main/java/domain/LottoMachine.java:
```diff
@@ -0,0 +1,47 @@
+        return numOfLotto;
+    }
+
+    private void checkNegativeInteger(int number) {
+        if (number < 0) {
+            throw new IllegalArgumentException(ERROR_COUNT_NEGATIVE_INTEGER);
+        }
+    }
```
음수 여부만 체크할 것이 아니라, 거스름돈 없이 로또를 살 수 있는 금액인지도 체크해야 하지 않을까요? 지금은 1500원을 입력해도 특별한 메시지 없이 그냥 하나만 살 수 있네요.

Seongwon97 replied:
1500원, 1700원과 같을 때는 돈은 자동 반환된다고 생각을 하고, 구매할 수 있는 최대한의 수로 로또 구매를 하도록 하였었습니다.(수익률 계산은 구매하는데 들어가는 돈으로만 계산)

피드백을 받아 해당 내용은 추가했습니다!

확실히 객체지향적으로 객체들을 분리하니까 Money클래스에 검증 하나만 추가하면 되어서 매우 편해진것 같네요😁

choihz commented on src/main/java/domain/LottoMachine.java:
```diff
@@ -0,0 +1,47 @@
+
+public class LottoMachine {
+
+    public static final int LOTTO_PRICE = 1000;
```
이 상수는 `LottoMachine`에 위치해 있지만 여기에서 쓰이진 않고, `Lottos`와 `Money`에서 쓰이고 있네요. 그렇다면 적절한 위치에 있는 걸까요?

Seongwon97 replied:
현재 로또 가격에 대한 정보를 `Lottos`와 `Money`에서 사용하고있고, 정의가 된 `LottoMachine`에서는 사용하지 않고 있습니다.

하지만 저는 객체지향적으로 생각하였을 때 로또 가격에 대한 정보는 `Lottos`와 `Money`보다는 로또를 판매하는 기기인 `LottoMachine`이 갖는 것이 더 좋다고 생각됩니다. 

만약 로또의 가격이 올라갈 경우, 자판기의 로또 가격 세팅을 변경하듯이 현재 프로그램에서도 `LottoMachine`이 해당 정보를 갖는 것이 더 좋다고 생각됩니다.
만약 다른 객체가 로또 가격의 정보를 갖는다면 `Lotto`가 가질 수 있을것 같습니다.(Lotto에서도 로또 가격 정보는 사용X)

choihz commented on src/main/java/util/InputUtil.java:
```diff
@@ -0,0 +1,23 @@
+        String[] splitNumbers = inputNumbers.split(LOTTO_NUMBER_DELIMITER);
+        return Arrays.asList(splitNumbers).stream()
+                .mapToInt(number -> validateAndChangeToInt(number, message))
+                .toArray();
```
```suggestion
        return Arrays.stream(splitNumbers)
                .mapToInt(number -> validateAndChangeToInt(number, message))
                .toArray();
```

choihz commented on src/main/java/util/InputUtil.java:
```diff
@@ -0,0 +1,23 @@
+
+import java.util.Arrays;
+
+public class InputUtil {
```
> 입력값에 대해 숫자로 변경가능한지 체크하고 정수로 변경하는 메서드, 그리고 로또본호 입력값에 대하여 split으로 분리를 하는 메서드는 프로그램 내에서 범용적으로 사용될 수 있다고 생각하여 util 클래스를 생성하여 해당 클래스 내에 위치시켰습니다.
> 해당 메서드들은 util성격에 맞다고 볼 수 있을까요??

아무래도 의견이 달라질 수 있는 부분인데요, `LOTTO_NUMBER_DELIMITER` 자체가 콘솔 뷰에 한정되는 정보이고, 클래스 이름도 이미 `InputUtil`로서 뷰에서만 사용될 것을 암시하고 있는 것 같아요. 그래서 **범용적**이라고 할 수 있는지는 모르겠습니다.

Seongwon97 replied:
![image](https://user-images.githubusercontent.com/57744251/156931068-9b873d5a-a45f-4252-9e4c-973ef6efe209.png)

그러면 코니는 이와 같이 뷰에서만 사용하도록 `util`디렉터리를 `view`안으로 넣는건 어떻게 생각하시나요? 

제 개인적인 생각으로는 `InputUtil`클래스의 메서드들은 적어도 해당 프로그램 내에서는 View안에서 많은 메서드에게 호출되어 범용적으로 쓰이고 있다고 생각합니다!!

choihz commented on src/main/java/domain/Rank.java:
```diff
@@ -37,12 +38,18 @@ public static Rank getWinnerPrizeByMatched(final int matched, final boolean bonu
+                .filter(winnerPrice -> winnerPrice != Rank.FAIL)
+                .sorted(Comparator.comparing(Rank::getPrize))
+                .collect(Collectors.toList());
+    }
```
이 메서드와 `getWinnerPrices()`는 이름만 보고 역할을 명확히 판단하기 어려운 것 같아요. 그리고 `getWinnerPrices()`는 `Rank`의 리스트를 리턴할 뿐인데 이름에 왜 **price**라는 단어가 들어간 걸까요?

Seongwon97 replied:
리펙터링 이전 클래스의 이름이 남아있던 것 같습니다..price가 아닌 rank라는 이름을 갖도록 수정했습니다!

choihz commented on src/main/java/domain/LottoMachine.java:
```diff
@@ -0,0 +1,47 @@
+import domain.generator.LottoGenerator;
+import domain.generator.ManualLottoGenerator;
+
+public class LottoMachine {
```
> 더욱 구체적인 LottoMachine은 어떻게 생각되시나요?? Store를 LottoMachine으로 변경해봤습니다.

이전 이름보다는 훨씬 나은 것 같습니다 👍

choihz commented on src/main/java/util/InputUtil.java:
```diff
@@ -0,0 +1,23 @@
+                .toArray();
+    }
+
+    public static int validateAndChangeToInt(final String number, final String message) {
```
> 변경된 메서드 명은 어떻게 생각하시나요? 저는 메서드의 내부 작업을 너무 많이 드러내는 것 같기는 한데..검증 뿐만 아니라 값 변환까지 해주는 것을 표현하여 이전의 이름보다는 더 좋은 것 같다고 생각됩니다👍🏻

사실 숫자로 변환한다는 의미만으로도 충분한 것 같긴 합니다 ㅎㅎ

choihz commented on src/main/java/domain/WinningNumbers.java:
```diff
@@ -1,38 +1,42 @@
 
     private final Set<Number> winningNumbers;
-    private Number bonus;
+    private final Number bonus;
```
사소하긴 하지만, `bonus`와 `bonusNumber`라는 이름이 클래스 전체에서 혼용되고 있네요. 하나로 통일해 주시면 더 좋을 것 같아요.

Seongwon97 replied:
보너스 번호에 대한 내용을 모두 `bonus`로 통일하였습니다!!

choihz commented on src/main/java/domain/WinningNumbers.java:
```diff
@@ -1,38 +1,42 @@
     }
 
-    public void addBonusNumber(final int number) {
+    private Number validateBonusNumber(final int number) {
```
이 메서드 역시 이름이 하는 일을 제대로 드러내지 못하는 것 같아요 😢

Seongwon97 replied:
아래의 `validate~`의 메서드 명들을 모두 `check~`로 바꾸는 과정에서 해당 메서드 명도 `checkRangeAndDuplicate`로 변경하여 메소드 내부에서 동작하는 내용을 더 표현해봤습니다😁

choihz commented on src/test/java/domain/WinningNumberTest.java:
```diff
@@ -2,34 +2,50 @@
+
+    @DisplayName("입력된 번호가 6개가 아니면 예외가 발생한다.")
+    @ParameterizedTest(name = "{index} {displayName} numbers={0}")
+    @ValueSource(strings = {"12, 40, 43, 12,", "12, 4, 35, 2, 7, 5, 6"})
+    void checkNumOfNumbers_throwIllegalException(final String inputNumbers) {
+        int[] numbers = Arrays.stream(inputNumbers.split(LOTTO_NUMBER_DELIMITER))
+                .mapToInt(Integer::parseInt)
+                .toArray();
```
테스트 자체보다는 테스트에서 사용할 데이터를 준비하는 부분에 너무 눈길이 많이 가는 것 같아요. 다른 방법, 또는 개선할 수 있는 방법은 없을까요?

Seongwon97 replied:
피드백을 받고 해당 숫자를 직접 만드는 것이 아닌 `generator`클래스를 생성하여 번호를 랜덤하게 생성해볼까?라는 생각을 해봤습니다...
하지만 테스트 별로 필요한 입력 데이터가 달라서 `generator`를 생성할 경우 한번 사용하고 버리는 일회성 메서드가 될 것 같아..오히려 좋지 않은 방법이 될 것 같더군요😭

테스트 코드를 개선할 방법이 따로 있을까요...?? 아무리 생각해도 생각이 떠오르지 않네요😢
혹시 추천해주실만한 학습 자료가 있을까요?

choihz commented on src/main/java/domain/WinningNumbers.java:
```diff
@@ -1,38 +1,42 @@
     }
 
-    public void addBonusNumber(final int number) {
+    private Number validateBonusNumber(final int number) {
```
언제 `check`이고 언제 `validate`인가요?

Seongwon97 replied:
check는 확인용, validate는 검증용이라는 생각을 하며 코드 작성을 하였는데, 둘의 역할은 사실 같은 내용인 것 같습니다..😅
`check`, `validate`를 모두 `check`로 통일하였습니다.

choihz reviewed (changes requested):
안녕하세요, 렉스!

피드백을 잘 반영해 주셨네요 👍 남겨주신 코멘트까지 전부 확인했고, 마지막으로 몇 가지 의견과 질문을 더 남겨 보았어요. 이외에도 더 궁금한 점이나 의견이 있으시다면 언제든 코멘트 또는 DM 주세요~

Seongwon97 committed:
refactor: 메서드 내부 로직 축약

- 메서드 내부에 축약이 가능한 부분을 축약

Seongwon97 committed:
refactor: 메서드, 변수 이름 수정

- 메서드와 변수의 이름을 통일 & 하는 작업에 맞게 수정

Seongwon97 committed:
refactor: 메서드, 변수 이름 수정

- 메서드와 변수의 이름을 통일 & 하는 작업에 맞게 수정
- check와 validate를 모두 check로 통일

Seongwon97 committed:
refactor: 로또 담첨 결과 확인 메서드를 Rank로 이동

- 로또 당첨 결과 확인 Lotto에서 Rank로 이동

Seongwon97 committed:
refactor: 당첨번호 클래스에서 Lotto객체를 변수로 사용하도록 수정

- Set<Number>로 관리하던 당첨 번호를 Lotto객체를 사용하도록 변경

Seongwon97 committed:
fix: 투입 금액이 1000원 단위가 아닌 경우에도 정상 실행되는 에러 수정

- 투입 금액이 1000의 배수가 아닐 경우 예외처리 추가

Seongwon97 committed:
refactor: utils 디렉터리 위치를 view아래로 변경

Seongwon97 committed:
refactor: 유효한 rank정보를 반환하는 메서드 이름 수정

- Rank.java의 유효한 당첨 정보를 반환하는 getValidPrizes메서드의 이름을
getValidRanks로 변경

Seongwon97 committed:
fix: 테스트 코드 input값 수정

Seongwon97 committed:
style: 코드 컨벤션 수정

Seongwon97 commented:
안녕하세요 코니

보내주신 피드백 내용들을 반영해보았어요!
아직 부족함이 많지만 코니의 피드백을 받으면 받을수록 코드가 더욱 객체지향적으로 분리되는 것을 크게 느끼고 있어요.
각각의 객체가 자기 자신이 해야할 일에 대한 정보와 메서드만 갖고있다보니 코드의 리펙터링도 처음보다는 점점 쉬워지는 것 같네요😁

오늘이 미션 마감일이라 한번의 피드백을 더 받을 수도, 받지 못할 수도 있지만
2주동안 좋은 피드백들을 많이 남겨주셔서 감사합니다🙇🏻‍♂️

Seongwon97 committed:
refactor: 변수, 메서드 명 수정

- winnerPrize -> winnerRank
- totalPrice -> totalPrize

choihz commented on src/main/java/domain/WinningNumbers.java:
```diff
@@ -1,46 +1,34 @@
-        final Number bonusNumber = new Number(number);
-        if (winningNumbers.contains(bonusNumber)) {
+    private Number checkBonusRangeAndDuplicate(final int number) {
+        final Number bonus = new Number(number);
+        if (winningNumbers.hasMatchedNumber(bonus)) {
             throw new IllegalArgumentException(ERROR_DUPLICATE_BONUS);
         }
-        bonus = bonusNumber;
+        return bonus;
     }
```
`Number` 객체 생성을 통해 숫자 범위를 검증하게 되기 때문에 메서드 이름에도 해당 내용이 포함된 걸까요? 구현을 너무 드러내는 이름 같아요 😢

choihz commented on src/main/java/domain/LottoMachine.java:
```diff
@@ -0,0 +1,47 @@
+
+public class LottoMachine {
+
+    public static final int LOTTO_PRICE = 1000;
```
> 현재 로또 가격에 대한 정보를 `Lottos`와 `Money`에서 사용하고있고, 정의가 된 `LottoMachine`에서는 사용하지 않고 있습니다.
> 
> 하지만 저는 객체지향적으로 생각하였을 때 로또 가격에 대한 정보는 `Lottos`와 `Money`보다는 로또를 판매하는 기기인 `LottoMachine`이 갖는 것이 더 좋다고 생각됩니다. 
> 
> 만약 로또의 가격이 올라갈 경우, 자판기의 로또 가격 세팅을 변경하듯이 현재 프로그램에서도 `LottoMachine`이 해당 정보를 갖는 것이 더 좋다고 생각됩니다.
> 만약 다른 객체가 로또 가격의 정보를 갖는다면 `Lotto`가 가질 수 있을것 같습니다.(Lotto에서도 로또 가격 정보는 사용X)

객체지향적으로 생각했을 때, 다른 객체의 퍼블릭 상수를 참조하면 불필요한 결합도는 증가하고 응집도는 낮아지는 측면이 있다고 말할 수 있습니다. 객체가 일하기 위해 알아야 할 정보를 스스로 알지 못하고 있는 거니까요. 따라서 이런 방식이 더 적절하게 느껴진다면, 여기서 설계를 개선할 포인트를 고민해 보면 좋을 것 같아요.

choihz commented on src/main/java/view/util/InputUtil.java:
```diff
@@ -0,0 +1,21 @@
+package view.util;
+
+import java.util.Arrays;
+
+public class InputUtil {
```
> 그러면 코니는 이와 같이 뷰에서만 사용하도록 `util`디렉터리를 `view`안으로 넣는건 어떻게 생각하시나요? 
> 
> 제 개인적인 생각으로는 `InputUtil`클래스의 메서드들은 적어도 해당 프로그램 내에서는 View안에서 많은 메서드에게 호출되어 범용적으로 쓰이고 있다고 생각합니다!!

개인적으로는, 최상단 하위의 `util`보다 `view` 하위의 `util`이 된 지금이 더 분명하고 적절한 것 같이 느껴져요 👍

choihz commented on src/test/java/domain/WinningNumberTest.java:
```diff
@@ -2,34 +2,50 @@
-        assertThatThrownBy(() -> lotto.addBonusNumber(bonus))
+    private static final String LOTTO_NUMBER_DELIMITER = ", |,";
+
+    @DisplayName("입력된 번호가 6개가 아니면 예외가 발생한다.")
+    @ParameterizedTest(name = "{index} {displayName} numbers={0}")
+    @ValueSource(strings = {"12, 40, 43, 12,", "12, 4, 35, 2, 7, 5, 6"})
+    void checkNumOfNumbers_throwIllegalException(final String inputNumbers) {
+        int[] numbers = Arrays.stream(inputNumbers.split(LOTTO_NUMBER_DELIMITER))
+                .mapToInt(Integer::parseInt)
+                .toArray();
```
> 피드백을 받고 해당 숫자를 직접 만드는 것이 아닌 `generator`클래스를 생성하여 번호를 랜덤하게 생성해볼까?라는 생각을 해봤습니다...
> 하지만 테스트 별로 필요한 입력 데이터가 달라서 `generator`를 생성할 경우 한번 사용하고 버리는 일회성 메서드가 될 것 같아..오히려 좋지 않은 방법이 될 것 같더군요😭
> 
> 테스트 코드를 개선할 방법이 따로 있을까요...?? 아무리 생각해도 생각이 떠오르지 않네요😢
> 혹시 추천해주실만한 학습 자료가 있을까요?

앗 어렵게 생각하실 필요는 없어요! 지금 상황에서는 공통 부분을 메서드로 분리하고 이름을 잘 지어주는 것만으로도 가독성이 훨씬 좋아질 것 같아요. 앞으로 과정을 진행하면서 경험을 통해 많이 배우실 수 있는 영역이라고 생각합니다 😉

choihz commented on src/test/java/domain/WinningNumberTest.java:
```diff
@@ -2,34 +2,50 @@
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessageContaining("보너스 번호가 당첨 번호와 중복됩니다.");
+    }
+}
```
미션 PR에 걸려있는 링크를 통해 [우아한테크코스 - 레벨1 로또 미션 피드백 정리](https://seongwon97.github.io/posts/%EC%9A%B0%ED%85%8C%EC%BD%94-%EB%A0%88%EB%B2%A81-%EB%A1%9C%EB%98%90-%ED%94%BC%EB%93%9C%EB%B0%B1%EC%A0%95%EB%A6%AC/)를 읽었어요. 정리를 아주 깔끔하게 잘 해주셨고 내용도 좋네요 👍 이렇게 정리한 것들은 나중에 정말 큰 자산이 될 거예요. 앞으로도 종종 구경갈게요 😊

choihz reviewed (approved):
안녕하세요, 렉스!

수정하신 부분과 질문주신 내용을 확인하고, 코멘트를 몇 개 달아 두었습니다. 이제 머지해도 될 것 같아요. 그동안 정말 고생 많으셨고, 다음에 다시 만나요~ 👋
